File: clojure/protein-translation/src/protein_translation.clj
(ns protein-translation)

(def acid->codons {"Methionine" ["AUG"]
                   "Phenylalanine" ["UUU" "UUC"]
                   "Leucine" ["UUA" "UUG"]
                   "Serine" ["UCU" "UCC" "UCA" "UCG"]
                   "Tyrosine" ["UAU" "UAC"]
                   "Cysteine" ["UGU" "UGC"]
                   "Tryptophan" ["UGG"]
                   :stop ["UAA" "UAG" "UGA"]})

(def codon->acid (into {} (for [[acid codons] acid->codons
                                codon codons]
                            [codon acid])))

(defn translate-rna
  "Translates an RNA string into amino acids."
  [rna]
  (loop [acc [] remaining rna]
    (cond
      (empty? remaining) acc
      (< (count remaining) 3) (throw (IllegalArgumentException. "Invalid codon")) 
      :else (let [candidate (codon->acid (subs remaining 0 3))]
              (cond
                (nil? candidate) (throw (IllegalArgumentException. "Invalid codon"))
                (= :stop candidate) acc
                :else (recur (conj acc candidate) (subs remaining 3)))))))

File: python/protein-translation/protein_translation.py
"""Translate an RNA sequence into proteins"""

CODONS_TO_ACID = [
    [['AUG'], 'Methionine'],
    [['UUU', 'UUC'], 'Phenylalanine'],
    [['UUA', 'UUG'], 'Leucine'],
    [['UCU', 'UCC', 'UCA', 'UCG'], 'Serine'],
    [['UAU', 'UAC'], 'Tyrosine'],
    [['UGU', 'UGC'], 'Cysteine'],
    [['UGG'], 'Tryptophan'],
    [['UAA', 'UAG', 'UGA'], 'STOP']
]

CODON_TO_ACID = {
    codon: acid
    for codons, acid in CODONS_TO_ACID
    for codon in codons
}

def proteins(strand):    
    result = []
    for codon in [strand[index:index+3] for index in range(0, len(strand), 3)]:
        acid = CODON_TO_ACID[codon]
        if acid == 'STOP':
            break
        result.append(acid)
    return result
    